default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+    _rating_function()
+    _class_specific_filter()
+    _starting_guess()
+
+    For improved performance, maybe override _filling_algorithm() with a more specialised algorithm as well.
+    """
+    basename = "Default - Do not build"     # base design name
+    description = "Base class Ship type"    # design description
+    useful_part_classes = ALL_META_CLASSES  # only these parts are considered in the design process
+
+    filter_useful_parts = True              # removes any part not belonging to self.useful_part_classes
+    filter_inefficient_parts = False        # removes cost-inefficient parts (less capacity and less capacity/cost)
+
+    running_index = 1                       # running index used in the design name
+    running_index_needs_update = True       # To check if we need to update the running_index after a load.
+
+    def __init__(self):
+        """Make sure to call this constructor in each subclass."""
+        self.species = None         # name of the piloting species (string)
+        self.hull = None            # hull object (not hullname!)
+        self.partnames = []         # list of partnames (string)
+        self.parts = []             # list of actual part objects
+        self.attacks = {}           # {damage:count}
+        self.structure = 0
+        self.shields = 0
+        self.fuel = 0
+        self.speed = 0
+        self.stealth = 0
+        self.detection = 0
+        self.troops = 0
+        self.colonisation = -1      # -1 since 0 indicates an outpost (capacity = 0)
+        self.production_cost = 9999
+        self.production_time = 1
+        self.pid = -1               # planetID for checks on production cost if not LocationInvariant.
+        self.additional_specifications = AdditionalSpecifications()
+
+    def evaluate(self):
+        """ Return a rating for the design.
+
+        First, check if the minimum requirements are met. If so, return the _rating_function().
+        Otherwise, return a large negative number scaling with distance to the requirements.
+
+        :returns: float - rating of the current part/hull combo
+        """
+        self.update_stats()
+        ret_val = 0
+
+        # If we do not meet the requirements, we want to return a negative rating.
+        # However, we also need to make sure, that the closer we are to requirements,
+        # the better our rating is so the optimizing heuristic finds "the right way".
+        if self.fuel < self.additional_specifications.minimum_fuel:
+            ret_val += -1000 * (self.additional_specifications.minimum_fuel - self.fuel)
+        if self.speed < self.additional_specifications.minimum_speed:
+            ret_val += -1000 * (self.additional_specifications.minimum_speed - self.speed)
+        if self.structure < self.additional_specifications.minimum_structure:
+            ret_val += -1000 * (self.additional_specifications.minimum_structure - self.structure)
```

Comment:
I think the naming and use of the variables "additional_specifications.minimum_x" is quite self-explanatory. It merely states the minimum requirements for the design to be acceptable.

For example, the AI in the future might ask to build a troop ship to invade a planet. However, the ship will need 5 fuel for that. So it will only consider designs with that amount of fuel. With the implementation I used, the AI is capable of seeing "oh wait, I need more fuel, let's throw out troop pods for some fuel tanks".

As for why I chose the numbers, refer to the existing comment.

For what you are suggesting: This is covered and handled by the _rating_function(). I suppose you will read up on that soon so I will not explain this any further at this point.

If questions remain then, feel free to ask.